testing/test_gateway.py

```diff
@@ -525,3 +525,82 @@ def sendback(channel):
     if interleave_getstatus:
         print(gw.remote_status())
     assert ch.receive(timeout=0.5) == 1234
+
+
+def test_assert_main_thread_only(execmodel, makegateway):
+    if execmodel.backend != "main_thread_only":
+        pytest.skip("can only run with main_thread_only")
+
+    gw = makegateway(spec=f"execmodel={execmodel.backend}//popen")
+
+    try:
+        # Submit multiple remote_exec requests in quick succession and
+        # assert that all tasks execute in the main thread. It is
+        # necessary to call receive on each channel before the next
+        # remote_exec call, since the channel will raise an error if
+        # concurrent remote_exec requests are submitted as in
+        # test_main_thread_only_concurrent_remote_exec_deadlock.
+        for i in range(10):
+            ch = gw.remote_exec(
+                """
+                    import time, threading
+                    time.sleep(0.02)
+                    channel.send(threading.current_thread() is threading.main_thread())
+            """
+            )
+
+            try:
+                res = ch.receive()
+            finally:
+                ch.close()
+                # This doesn't actually block because we closed
+                # the channel already, but it does check for remote
+                # errors and raise them.
+                ch.waitclose()
```

Comment:
The channel close and waitclose methods do not actually elicit a response from the remote end, so at this point in time, in theory the remote main thread could still be busy and therefore it could be unsafe to send the next `remote_exec` call at this moment in time. The earlier receive call doesn't really provide a hard guarantee either because the message was sent from within the main thread (also the WorkerGateway executetask method needs to finish executing in the main thread).

It bothers me that we don't have a hard guarantee here, but at the same time the probability of triggering the deadlock error here seems extremely close to zero. For pytest-xdist it doesn't even matter because it never makes more than a single `remote_exec` call per channel.

I suppose we could disallow more than a single `remote_exec` call per channel on the client side. What do you think @RonnyPfannschmidt?

Reply:
That seems pretty reasonable but also kind of a sad way to present this thread model.

I've just realize that if we add a timeout before we trigger the deadlock error in the _local_schedulexec method, then that will give us a hard guarantee of correctness in practice here. We can add use a threading.Event instance in WorkerGateway and wait for the tail of the executetask method to set this event after there is no longer any possibility of a deadlock. If we wait on this event for about 1 second before we resort to raising the deadlock error, then it's practically impossible for it to be a false positive.

Reply:
If it doesn't add too much complexity we can try

Long term goal is a sans async event loop in a thread

Reply:
It came out pretty simple, and allows us to avoid messing around with self._execpool._primary_thread_task_ready:
```diff
--- a/src/execnet/gateway_base.py
+++ b/src/execnet/gateway_base.py
@@ -1146,9 +1146,14 @@ class WorkerGateway(BaseGateway):
         if self._execpool.execmodel.backend == "main_thread_only":
-            primary_thread_task_ready = self._execpool._primary_thread_task_ready
-            if (
-                primary_thread_task_ready is not None
-                and primary_thread_task_ready.is_set()
-            ):
+            # It's necessary to wait for a short time in order to ensure
+            # that we do not report a false-positive deadlock error, since
+            # channel close does not elicit a response that would provide
+            # a guarantee to remote_exec callers that the previous task
+            # has released the main thread. If the timeout expires then it
+            # should be practically impossible to report a false-positive.
+            if not self._executetask_complete.wait(timeout=1):
                 channel.close(MAIN_THREAD_ONLY_DEADLOCK_TEXT)
                 return
+            # It's only safe to clear here because the above wait proves
+            # that there is not a previous task about to set it again.
+            self._executetask_complete.clear()
 
@@ -1183,2 +1188,8 @@ class WorkerGateway(BaseGateway):
         self._execpool = WorkerPool(self.execmodel, hasprimary=hasprimary)
+        self._executetask_complete = None
+        if self.execmodel.backend == "main_thread_only":
+            self._executetask_complete = self.execmodel.Event()
+            # Initialize state to indicate that there is no previous task
+            # executing so that we don't need a separate flag to track this.
+            self._executetask_complete.set()
         trace("spawning receiver thread")
@@ -1225,2 +1236,7 @@ class WorkerGateway(BaseGateway):
         channel.close()
+        if self._executetask_complete is not None:
+            # Indicate that this task has finished executing, meaning
+            # that there is no possibility of it triggering a deadlock
+            # for the next spawn call.
+            self._executetask_complete.set()
```